clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -777,6 +783,38 @@ constructKernelName(Sema &S, FunctionDecl *KernelCallerFunc,
                                       KernelNameType)};
 }
 
+static bool isDefaultSPIRArch(ASTContext &Context) {
+  llvm::Triple T = Context.getTargetInfo().getTriple();
+  if (T.isSPIR() && T.getSubArch() == llvm::Triple::NoSubArch)
+    return true;
+  return false;
+}
+
+static ParmVarDecl *getSyclKernelHandlerArg(FunctionDecl *KernelCallerFunc) {
+  // Specialization constants in SYCL 2020 are not captured by lambda and
+  // accessed through new optional lambda argument kernel_handler
+  auto KHArg =
+      std::find_if(KernelCallerFunc->param_begin(),
+                   KernelCallerFunc->param_end(), [](ParmVarDecl *PVD) {
+                     return Util::isSyclKernelHandlerType(PVD->getType());
+                   });
+
+  ParmVarDecl *KernelHandlerArg =
+      (KHArg != KernelCallerFunc->param_end()) ? *KHArg : nullptr;
+
+  if (KernelHandlerArg) {
```

Comment:
This whole block shouldn't appear in non-asserts mode.  What about removing 802 and on with:

assert(KHArg == llvm::find_if(llvm::reverse(KernelCallerFunc.parameters(), [](ParmVarDecl *PVD){ return ...; })); ?

Also, allows you to go back to the previous return statement ternary.